### DISCUSSION TITLE
How do we make it easy for viewers to reproduce non-cursorless commands in their own setup?

### DISCUSSION THREAD

**Discussion by pokey** [cursorless-dev/cursorless#547] [Q&A]
My voice commands have diverged significantly from knausj, meaning that it is not easy for a viewer to watch one of my videos and immediately start using the commands that they see

**Comment by pokey** [comment]
One option is for me to record my videos using stock knausj. Unfortunately, this isn't really an option for most of my videos, because my muscle memory is trained on a very different set of voice commands

However, I do think that it makes sense to do this for some videos. For example, it's probably a good idea to do this for the tutorial, or for shorter videos, where the focus is on a specific cursorless feature

**Comment by AndreasArvidsson** [comment]
Do a mapping between your words and the default knausj ones. Then in post processing just dub over them ;)

**Reply by pokey** [reply]
Haha that would be pretty hilarious

**Comment by pokey** [comment]
In some of my videos, I just make an effort to mention every time I use a custom command

I think this approach helps, but has its flaws, as

- I forget to do it sometimes
- It doesn't help users figure out how to add the given command to their own setup
- It slows down my videos for those that already know that command
- It can detract from the narrative flow of the video

**Comment by pokey** [comment]
I'm working on an automated voice command transcript generator that would activate when I'm recording a video. At the outset, it will just generate a separate webpage that users can go to for a given video, and it will have, for each command,

- a timestamp of where it appeared in the video
- A link to the talon file in my GitHub repo where it is defined, at the exact commit I was on when I recorded the video
- The rule that was used by the command

Note that the latter two will just be based on a hacky parse of the `speechsystem._sim` output

In the future we can get fancier by

- [x] Linking to the line of the voice command
- [ ] automatically generating a description of the command based on interpolated docstring. Like a dynamic version of https://github.com/wenkokke/talon-cheatsheet
- [x] embedding the video in a page so that the transcript can scroll alongside the video.
- [ ] Could even embed the whole thing in a webview in VSCode, and allow them to pull up the file I was hacking on to practice on

**Comment by pokey** [comment]
One possibility is to make it easier for users to use something much closer to my talon files. I think the easiest way to get here is to improve the ability to customise knausj without forking it

To this end, I'd like to take full inventory of the changes on my fork, and decide, for each one, whether to upstream it, or to change knausj so that I can customise it without directly changing it. There are a couple techniques that could help here:

- Introducing the csv spoken form customisation mechanism supported by cursorless to knausj
- Doing something like https://github.com/knausj85/knausj_talon/pull/758 for situations where we want to customise behaviour

**Reply by david-tejada** [reply]
+1 to introducing csv customization to talon

**Comment by pokey** [comment]
Have more detailed / smarter subtitles and history embedded in the video, which include descriptions

**Comment by RonWalker22** [comment]
It would be even easier for the user if there was not so much focus on the defaults. What about those who don’t use defaults for everything? When you produce all the educational material bounded to the defaults it gives another reason for new people not to change the defaults. And sometimes they are forced to change because of enunciation/microphone quality issues. 

Along with presenter’s customizations, you could map the user’s customizations too and that would allow you to bypass and or supplement the defaults. To do the mapping maybe the user can share their fork or a json file that lists their customizations.
